hilti/toolchain/src/compiler/cfg.cc

```diff
@@ -0,0 +1,912 @@
+// Copyright (c) 2020-now by the Zeek Project. See LICENSE for details.
+
+#include "hilti/compiler/detail/cfg.h"
+
+#include <algorithm>
+#include <cassert>
+#include <cstdint>
+#include <iterator>
+#include <map>
+#include <optional>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include <hilti/rt/util.h>
+
+#include <hilti/ast/attribute.h>
+#include <hilti/ast/ctors/tuple.h>
+#include <hilti/ast/declaration.h>
+#include <hilti/ast/declarations/expression.h>
+#include <hilti/ast/declarations/function.h>
+#include <hilti/ast/declarations/global-variable.h>
+#include <hilti/ast/declarations/local-variable.h>
+#include <hilti/ast/declarations/module.h>
+#include <hilti/ast/declarations/parameter.h>
+#include <hilti/ast/expression.h>
+#include <hilti/ast/expressions/assign.h>
+#include <hilti/ast/expressions/ctor.h>
+#include <hilti/ast/expressions/keyword.h>
+#include <hilti/ast/expressions/member.h>
+#include <hilti/ast/expressions/name.h>
+#include <hilti/ast/expressions/resolved-operator.h>
+#include <hilti/ast/location.h>
+#include <hilti/ast/node.h>
+#include <hilti/ast/operator.h>
+#include <hilti/ast/operators/function.h>
+#include <hilti/ast/operators/struct.h>
+#include <hilti/ast/scope-lookup.h>
+#include <hilti/ast/statement.h>
+#include <hilti/ast/statements/assert.h>
+#include <hilti/ast/statements/block.h>
+#include <hilti/ast/statements/break.h>
+#include <hilti/ast/statements/comment.h>
+#include <hilti/ast/statements/continue.h>
+#include <hilti/ast/statements/declaration.h>
+#include <hilti/ast/statements/expression.h>
+#include <hilti/ast/statements/for.h>
+#include <hilti/ast/statements/if.h>
+#include <hilti/ast/statements/return.h>
+#include <hilti/ast/statements/set_location.h>
+#include <hilti/ast/statements/throw.h>
+#include <hilti/ast/statements/try.h>
+#include <hilti/ast/statements/while.h>
+#include <hilti/ast/statements/yield.h>
+#include <hilti/ast/type.h>
+#include <hilti/ast/types/reference.h>
+#include <hilti/ast/types/struct.h>
+#include <hilti/ast/visitor.h>
+#include <hilti/base/util.h>
+#include <hilti/hilti/ast/types/bytes.h>
+#include <hilti/hilti/ast/types/list.h>
+#include <hilti/hilti/ast/types/map.h>
+#include <hilti/hilti/ast/types/set.h>
+#include <hilti/hilti/ast/types/stream.h>
+#include <hilti/hilti/ast/types/vector.h>
+
+namespace hilti::detail::cfg {
+
+// Helper function to detect whether values of a type can alias their inputs.
+static bool _isAliasingType(const UnqualifiedType& type) {
+    // TODO(bbannier): Make this part of e.g., `UnqualifiedType` instead of a hardcoded list of types here?
```

Comment:
(Just as a note, moving this into `UnqualifiedType` makes sense to me. But can do later.)